File: app/models/well.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: wells
#
#  id           :integer          not null, primary key
#  sample_id    :integer
#  wellplate_id :integer          not null
#  position_x   :integer
#  position_y   :integer
#  created_at   :datetime         not null
#  updated_at   :datetime         not null
#  additive     :string
#  deleted_at   :datetime
#  readouts     :jsonb
#  label        :string           default("Molecular structure"), not null
#  color_code   :string
#
# Indexes
#
#  index_wells_on_deleted_at    (deleted_at)
#  index_wells_on_sample_id     (sample_id)
#  index_wells_on_wellplate_id  (wellplate_id)
#

class Well < ApplicationRecord
  HEX_COLOR_REGEX = /\A#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})\z/.freeze
  # Mirrors the +wells.label+ column default; a well still carrying it has not
  # been labelled by a user. See {#content?}.
  DEFAULT_LABEL = 'Molecular structure'
  has_logidze
  acts_as_paranoid
  belongs_to :wellplate
  belongs_to :sample, optional: true

  include Tagging

  validates :color_code,
            format: { with: HEX_COLOR_REGEX, message: :invalid_hex_color },
            allow_blank: true

  def self.get_samples_in_wellplates(wellplate_ids)
    where(wellplate_id: wellplate_ids).pluck(:sample_id).compact.uniq
  end

  def readouts
    read_attribute(:readouts) || []
  end

  # Whether the well holds anything a user put there.
  #
  # Used to decide whether shrinking a wellplate may drop this well. Both
  # +label+ and +readouts+ have non-null column defaults, so a bare presence
  # check would report every untouched well as occupied and make shrinking
  # impossible. +sample+ is read through the association rather than
  # +sample_id+ on purpose: {Sample} is +acts_as_paranoid+ and
  # {Usecases::Wellplates::WellplateUpdater#update_wells} destroys samples
  # without nulling the column, so dangling ids are common and must read as
  # empty.
  #
  # @return [Boolean]
  def content?
    sample.present? || additive.present? || color_code.present? ||
      (label.present? && label != DEFAULT_LABEL) ||
      readouts.any? { |readout| readout_filled?(readout) }
  end

  # translates well position within wellplate: X=2 Y=3 -> C2
  def alphanumeric_position
    # A non-positive row would index the label table from the end (y=0 -> "ZZ"),
    # naming a cell that does not exist. Reachable since Resize reports wells
    # with a below-one position as blocking a shrink.
    return 'n/a' unless placeable?

    row = ('A'..'ZZ').to_a[position_y - 1]
    "#{row}#{position_x}"
  end

  def sortable_alphanumeric_position
    return 'n/a' unless placeable?

    row = ('A'..'ZZ').to_a[position_y - 1]
    "#{row}#{format('%02i', position_x)}"
  end

  # Whether the well sits at a position any grid could hold. Mirrors
  # Wellplate.positionOutside on the client.
  #
  # @return [Boolean]
  def placeable?
    position_x.present? && position_y.present? && position_x.positive? && position_y.positive?
  end

  private

  # @param readout [Object] one entry of {#readouts}; jsonb yields string keys
  #   when loaded from the DB and symbol keys when built in memory
  # @return [Boolean] whether the entry carries a value or a unit
  def readout_filled?(readout)
    return false unless readout.is_a?(Hash)

    entry = readout.with_indifferent_access
    entry[:value].present? || entry[:unit].present?
  end
end

File: app/javascript/src/models/Well.js
/* eslint-disable no-underscore-dangle */
import Element from 'src/models/Element';
import Sample from 'src/models/Sample';
import Wellplate from 'src/models/Wellplate';

// Mirrors ActiveSupport's `.present?`, which the server's Well#content? uses.
// A plain truthiness test disagrees with it on a numeric 0 - and a spreadsheet
// import stores raw cell values, so `{ value: 0 }` is a real readout. The
// server would call that well occupied and refuse a shrink while the client
// called it empty and offered the size as safe.
const isPresent = (value) => value !== null && value !== undefined && String(value).trim() !== '';

export default class Well extends Element {
  // Mirrors the wells.label column default; see `hasContent`.
  static get DEFAULT_LABEL() {
    return 'Molecular structure';
  }

  /**
   * Whether a position is one some grid could actually hold. A non-positive
   * row is not: rowLabel(0) is '' here and indexes the label table from the end
   * on the server, so the same well would be named two different
   * non-existent cells.
   */
  static isPlaceable(position) {
    if (!position || position.x == null || position.y == null) return false;

    return position.x > 0 && position.y > 0;
  }

  serialize() {
    return super.serialize({
      position: this.position,
      readouts: this.readouts || [],
      sample: this.sample && new Sample(this.sample).serialize(),
      color_code: this.color_code,
      label: this.label
    });
  }

  set sample(sample) {
    this._sample = (sample) ? new Sample(sample) : null;
  }

  get sample() {
    return this._sample;
  }

  get color_code() {
    return this._color_code;
  }

  set color_code(colorCode) {
    this._color_code = colorCode;
  }

  get label() {
    return this._label;
  }

  set label(label) {
    this._label = label;
  }

  /**
   * Whether the well holds anything a user put there.
   *
   * Mirrors Well#content? on the server, which is the authority. Both `label`
   * and `readouts` have non-null column defaults, so a bare presence check
   * would report every untouched well as occupied and make shrinking a
   * wellplate impossible.
   *
   * `additive`, `color_code` and `label` are exposed at detail level 10, so
   * below that WellEntity hands back the '***' placeholder for them. Counting
   * a placeholder as content would mark every well of such a plate occupied -
   * `label` alone would do it - and grey out every smaller size.
   */
  get hasContent() {
    if (this.sample) return true;
    if (this.additive && !this.isMethodDisabled('additive')) return true;
    if (this.color_code && !this.isMethodDisabled('color_code')) return true;
    if (this.label && this.label !== Well.DEFAULT_LABEL && !this.isMethodDisabled('label')) return true;

    // Array.isArray, not a truthiness check: readouts is exposed at
    // anonymize_below 1, so below that detail level the entity hands back the
    // '***' placeholder string and `.some` would not exist. Any non-array is
    // treated as "no readouts we can read", which is the safe reading.
    if (!Array.isArray(this.readouts)) return false;

    return this.readouts.some((readout) => readout && (isPresent(readout.value) || isPresent(readout.unit)));
  }

  get alphanumericPosition() {
    // Mirrors Well#alphanumeric_position on the server. A well with no usable
    // position is reachable here: positionOutside deliberately counts one as
    // outside any grid, so it can reach the "these wells block the resize"
    // message. Without this, rowLabel(null) is undefined and the position
    // renders as NaN - or throws outright when position itself is absent.
    if (!Well.isPlaceable(this.position)) return 'n/a';

    const positionY = Wellplate.rowLabel(this.position.y)
    const position = positionY + this.position.x;

    return position
  }
}
